examples/notification-demo/echo_server/echo.c

```diff
@@ -0,0 +1,1975 @@
+    }
+
+    return 0;
+}
+
+static int http_write_header(WOLFSSL *ssl, char *http_buf, size_t http_buf_size, int http_code, ssize_t content_length, const char *content_type, const char *body) {
+    time_t now = time(NULL);
+    struct tm now_tm;
+    char now_formatted[32];
+    size_t now_formatted_len;
+    int i;
+    const char *http_code_description = NULL, *http_code_explanation = NULL;
+    int resp_len, ret;
+
+    static const struct { int code; char *description; char *explanation; } http_codes[] = {
+        { 200, "OK", "Operation succeeded\n" },
+        { 403, "Forbidden", "Authorization denied.\n" },
+        { 404, "Not Found", "Requested path not recognized.\n" },
+        { 500, "Internal Server Error", "Internal server error.\n" }
+    };
+
+    for (i = 0; i < (int)(sizeof http_codes / sizeof http_codes[0]); ++i) {
+        if (http_codes[i].code == http_code) {
+            http_code_description = http_codes[i].description;
+            http_code_explanation = http_codes[i].explanation;
+            break;
+        }
+    }
+    if (http_code_description == NULL)
+        http_code_description = "Unknown code";
+
+    if (gmtime_r(&now, &now_tm) == NULL) {
+        fprintf(stderr, "gmtime_r failed\n");
+        return -1;
+    }
+
+    now_formatted_len = strftime(now_formatted, sizeof now_formatted, "%a, %d %h %Y %H:%M:%S %Z", &now_tm);
+    if (now_formatted_len == 0) {
+        fprintf(stderr, "strftime failed\n");
+        return -1;
+    }
+
+    if ((content_length < 0) && body)
+        content_length = strlen(body);
+
+    if ((content_length == 0) && http_code_explanation) {
+        content_type = "text/plain";
+        content_length = strlen(http_code_explanation);
+        body = http_code_explanation;
+    } else if (! body)
+        body = "";
+
+    resp_len = snprintf(http_buf, http_buf_size, "HTTP/1.0 %d %s\r\n" \
+        "Date: %.*s\r\nServer: Helicopter\r\nContent-Length: %zu%s%s\r\n\r\n%s",
+                        http_code,
+                        http_code_description,
+                        (int)now_formatted_len,
+                        now_formatted,
+                        content_length,
+                        content_type ? "\r\nContent-Type: " : "",
+                        content_type ? content_type : "",
+                        body);
+    if (resp_len > (int)http_buf_size) {
+        fprintf(stderr, "overrun averted while formatting header, " \
+            "http_buf_size = %zu, needed = %d\n", http_buf_size, resp_len);
+        return -1;
+    }
+    if ((ret = wolfSSL_write(ssl, http_buf, resp_len)) != resp_len) {
+        fprintf(stderr, "ERROR: failed to write: %s\n",
+            wolfSSL_ERR_reason_error_string(wolfSSL_get_error(ssl, ret)));
+        return -1;
+    }
+    return 0;
+}
+
+#ifndef NO_FILESYSTEM
+/* reads file size, allocates buffer, reads into buffer, returns buffer */
+#include <stdio.h>
+static int load_file(const char* fname, char** buf, size_t* bufLen)
```

Comment:
didn't want to use the `mmap()` technique for this?

Reply:
It is not portable.